docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+and respond with an error to any communications from older generations: this would help isolated
+pageserver nodes to shut down more quickly once they realize that they are stale, rather than continuing
+to put load on the safekeeper and write objects to S3 that are already superseded by later generations.
+
+For strict safety, the safekeeper should ensure that it does not trim any data based on messages
+from stale pageserver generations: this would require a series of barriers, where the control plane
+tells safekeepers about a new generation ID before it issues it to a pageserver. Implementing this
+is a lower priority than the underlying format changes, because the lifetime of split-brain situations
+is likely to be very short compared with the length of data retention in the safekeeper.
+
+Implementation of safekeeper changes is optional as long as the safekeeper keeps tenant WAL data
+indefinitely. Once we start deleting that, it becomes necessary for the safekeeper to only
+obey updates to the remote persistent LSN from nodes whose generation numbers are not stale.
+
+# Appendix A: Examples of use in high availability/failover
+
+The generation numbers proposed in this RFC are adaptable to a variety of different
+failover scenarios and models. The sections below sketch how they would work in practice.
+
+### Fast restart of a pageserver
+
+"fast" here means that the restart is done before any other element in the system
+has taken action in response to the node being down.
+
+- After restart, the node does no writes until it can obtain a fresh generation
+  number from the control plane.
+- Once issued a new generation number, the node synchronizes with control plane to learn of any
+  changes to its attachments, and receives a fast O(1) response informing it that it may continue
+  to operate with all its existing attachments.
+- Node proceeds to operate as normal, trusting that all its local content
+  for attached tenants is up to date.
+
+The only difference between this and how the system behaves today is that
+the node generation number increases, and the pageserver's startup depends
+on availability of the control plane. The improvement is an increase in _safety_: we
+are no longer assuming that we may continue to write to all our attached
+tenants' data, we are confirming it before writing.
+
+### Unexpected "death" (possible network partition) of a pageserver
+
+Consider a network partition where somewhere there is a functioning node
+that believes it is the rightful holder of a particular node ID, and it
+can write to S3 even though it can't communicate with our control plane.
+
+This is the general case of a node dying: we must always assume that the
+unresponsive node might be network partitioned and still writing to S3.
+
+#### Case A: re-attachment to other nodes
+
+1. Let's say node 0 fails in a cluster of three nodes 0, 1, 2.
```

Comment:
NIT: all paragraphs have `1.` seqno

Reply:
This is valid markdown (it generates the numbers when rendering), but I'll make it actual numbers so that the doc is easier to read as plain text.